File: design-pattern/factoryMethod/abstract_factory.ts
interface PointGuard {
    assist(): void
}

class Paul implements PointGuard {
    assist(): void {
        console.log('Paul assist')
    }
}

class Nash implements PointGuard {
    assist(): void {
        console.log('Nash assist')
    }
}

interface CentreForward {
    slamDunk(): void
}

class ONeal implements CentreForward {
    slamDunk(): void {
        console.log('ONeal slam dunk')
    }
}

class YaoMing implements CentreForward {
    slamDunk(): void {
        console.log('YaoMing slam dunk')
    }
}

interface TeamFactory {
    createPointGuard(): PointGuard

    createCentreForward(): CentreForward
}

class RocketTeam implements TeamFactory {
    createCentreForward(): CentreForward {
        return new YaoMing();
    }

    createPointGuard(): PointGuard {
        return new Paul();
    }
}

class LakersTeam implements TeamFactory {
    createCentreForward(): CentreForward {
        return new ONeal();
    }

    createPointGuard(): PointGuard {
        return new Nash();
    }
}

function play(teamFactory: TeamFactory) {
    const centreForward = teamFactory.createCentreForward()
    const pointGuard = teamFactory.createPointGuard()
    pointGuard.assist()
    centreForward.slamDunk()
}

play(new RocketTeam())
play(new LakersTeam())

File: design-pattern/factoryMethod/abstract_factory.rs
pub trait PointGuard {
    fn assist(&self);
}

pub struct Paul;

impl PointGuard for Paul {
    fn assist(&self) {
        println!("Paul assist");
    }
}

struct Nash;

impl PointGuard for Nash {
    fn assist(&self) {
        println!("Nash assist");
    }
}

pub trait CentreForward {
    fn slam_dunk(&self);
}

pub struct ONeal;

impl CentreForward for ONeal {
    fn slam_dunk(&self) {
        println!("ONeal slam dunk");
    }
}

struct YaoMing;

impl CentreForward for YaoMing {
    fn slam_dunk(&self) {
        println!("YaoMing slam dunk");
    }
}

pub trait TeamFactory {
    fn create_point_guard(&self) -> Box<dyn PointGuard>;
    fn create_centre_forward(&self) -> Box<dyn CentreForward>;
}

pub struct LakersFactory;

impl TeamFactory for LakersFactory {
    fn create_point_guard(&self) -> Box<dyn PointGuard> {
        Box::new(Paul)
    }

    fn create_centre_forward(&self) -> Box<dyn CentreForward> {
        Box::new(ONeal)
    }
}

pub struct RocketFactory;

impl TeamFactory for RocketFactory {
    fn create_point_guard(&self) -> Box<dyn PointGuard> {
        Box::new(Nash)
    }

    fn create_centre_forward(&self) -> Box<dyn CentreForward> {
        Box::new(YaoMing)
    }
}
